src/backend/dx11/src/device.rs

```diff
@@ -1,983 +1,1322 @@
+            )
         };
-        if !winapi::SUCCEEDED(hr) {
-            error!("Failed to create SRV from {:#?}, error {:x}", native_desc, hr);
-            return Err(d::ResourceViewError::Unsupported);
+
+        assert_eq!(true, winerror::SUCCEEDED(hr));
+
+        Sampler {
+            sampler_handle: unsafe { ComPtr::from_raw(sampler) }
         }
-        Ok(self.share.handles.borrow_mut().make_texture_srv(native::Srv(raw_view), htex))
     }
 
-    fn view_texture_as_unordered_access_raw(&mut self, _htex: &h::RawTexture<R>)
-                                        -> Result<h::RawUnorderedAccessView<R>, d::ResourceViewError> {
-        Err(d::ResourceViewError::Unsupported) //TODO
+    fn create_descriptor_pool<I>(
+        &self,
+        max_sets: usize,
+        descriptor_pools: I,
+    ) -> DescriptorPool
+    where
+        I: IntoIterator,
+        I::Item: Borrow<pso::DescriptorRangeDesc>
+    {
+        // TODO: descriptor pool
+
+        DescriptorPool
     }
 
-    fn view_texture_as_render_target_raw(&mut self, htex: &h::RawTexture<R>, desc: texture::RenderDesc)
-                                         -> Result<h::RawRenderTargetView<R>, d::TargetViewError>
+    fn create_descriptor_set_layout<I, J>(
+        &self, bindings: I, _immutable_samplers: J
+    ) -> DescriptorSetLayout
+    where
+        I: IntoIterator,
+        I::Item: Borrow<pso::DescriptorSetLayoutBinding>,
+        J: IntoIterator,
+        J::Item: Borrow<Sampler>,
     {
-        use winapi::UINT;
-        use core::texture::{AaMode, Kind};
-        use data::map_format;
-
-        let level = desc.level as UINT;
-        let (dim, extra) = match (htex.get_info().kind, desc.layer) {
-            (Kind::D1(..), None) =>
-                (winapi::D3D11_RTV_DIMENSION_TEXTURE1D, [level, 0, 0]),
-            (Kind::D1Array(_, nlayers), Some(lid)) if lid < nlayers =>
-                (winapi::D3D11_RTV_DIMENSION_TEXTURE1DARRAY, [level, lid as UINT, 1+lid as UINT]),
-            (Kind::D1Array(_, nlayers), None) =>
-                (winapi::D3D11_RTV_DIMENSION_TEXTURE1DARRAY, [level, 0, nlayers as UINT]),
-            (Kind::D2(_, _, AaMode::Single), None) =>
-                (winapi::D3D11_RTV_DIMENSION_TEXTURE2D, [level, 0, 0]),
-            (Kind::D2(_, _, _), None) if level == 0 =>
-                (winapi::D3D11_RTV_DIMENSION_TEXTURE2DMS, [0, 0, 0]),
-            (Kind::D2Array(_, _, nlayers, AaMode::Single), None) =>
-                (winapi::D3D11_RTV_DIMENSION_TEXTURE2DARRAY, [level, 0, nlayers as UINT]),
-            (Kind::D2Array(_, _, nlayers, AaMode::Single), Some(lid)) if lid < nlayers =>
-                (winapi::D3D11_RTV_DIMENSION_TEXTURE2DARRAY, [level, lid as UINT, 1+lid as UINT]),
-            (Kind::D2Array(_, _, nlayers, _), None) if level == 0 =>
-                (winapi::D3D11_RTV_DIMENSION_TEXTURE2DMSARRAY, [0, nlayers as UINT, 0]),
-            (Kind::D2Array(_, _, nlayers, _), Some(lid)) if level == 0 && lid
```

Comment:
similarly, please use `log` crate macros

Reply:
Did you comment on the right place? this section is from the old stuff in the file which got nuked, same for two comments below